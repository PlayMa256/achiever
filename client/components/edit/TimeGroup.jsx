import React from 'react';
import PropTypes from 'prop-types';

import TimeField from './TimeField';

import './TimeGroup.styl';

const getClassName = props =>
	`time-group ${props.emphasis ? 'emphasis' : ''}`;

export default class TimeGroup extends React.Component {
	constructor(props) {
		super(props);

		this.randomId = String(Math.random());

		this.onChangeTime = this.onChangeTime.bind(this);
	}

	onChangeTime(mode) {
		return (dummy, suggestion) => {
			const newtime = { ...this.props.time };
			newtime[mode] = suggestion;
			this.props.onSet(newtime.hours, newtime.minutes);
		};
	}

	render() {
		const {
			referenceHour,
			time,
			label,
			hidden,
			disabled
		} = this.props;

		if (hidden) {
			return null;
		}
		return (
			<div className={getClassName(this.props)}>
				<label htmlFor={`${this.randomId}_h`}>{label}</label>
				<TimeField
					mode="hours"
					value={time.hours}
					referenceHour={referenceHour}
					shouldHaveFocus={this.props.shouldHaveFocus === 'hours'}
					onChange={this.onChangeTime('hours')}
					onFocus={this.props.onFocus}
					tabIndex={this.props.tabIndexes}
					disabled={disabled}
					handleKeyPress={this.props.handleKeyPress}
				/>
				<span className="separator">:</span>
				<TimeField
					mode="minutes"
					value={time.minutes}
					shouldHaveFocus={this.props.shouldHaveFocus === 'minutes'}
					onChange={this.onChangeTime('minutes')}
					onFocus={this.props.onFocus}
					tabIndex={this.props.tabIndexes + 1}
					disabled={disabled}
					handleKeyPress={this.props.handleKeyPress}
				/>
			</div>
		);
	}
}

TimeGroup.propTypes = {
	onSet: PropTypes.func,
	referenceHour: PropTypes.number,
	time: PropTypes.object,
	label: PropTypes.string,
	shouldHaveFocus: PropTypes.oneOfType([
		PropTypes.oneOf(['hours', 'minutes']),
		PropTypes.bool
	]),
	onFocus: PropTypes.func.isRequired,
	tabIndexes: PropTypes.number,
	hidden: PropTypes.bool,
	disabled: PropTypes.bool,
	handleKeyPress: PropTypes.func
};

TimeGroup.defaultProps = {
	referenceHour: 9,
	label: '',
	shouldHaveFocus: false,
	time: { hours: 0, minutes: 0 },
	tabIndexes: 0,
	hidden: false,
	disabled: false,
	onSet: () => {},
	handleKeyPress: () => {}
};
