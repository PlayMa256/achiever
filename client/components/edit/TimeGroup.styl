@import "_colours.styl"
@import "_standards.styl"
	
.time-group
	min-height: 70px;
	clear: both;

	label
		label-style()

		.emphasis &
			font-size: 13px
	
	.separator
		display: block;
		float: left;
		font-size: 24px;
		width: 10px;
		text-align: center;
		margin: 10px 8px;

	input
		&.minutes, &.hours
			display: block;
			float: left;
			font-size: 30px;
			padding: 6px 12px;
			width: 102px;
			

.time-group input::placeholder
	color: fadedTextColour;

