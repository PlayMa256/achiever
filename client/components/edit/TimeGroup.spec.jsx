import React from 'react';
import { shallow } from 'enzyme';

import TimeGroup from './TimeGroup';

Math.random = jest.fn(() => '12345');

describe('TimeGroup', () => {
	describe('render', () => {
		it('should render two TimeFields', () => {
			const wrapper = shallow(<TimeGroup onFocus={() => {}} />);
			expect(wrapper).toMatchSnapshot();
		});
		it('should not render if hidden', () => {
			const wrapper = shallow(<TimeGroup onFocus={() => {}} hidden />);
			expect(wrapper).toMatchSnapshot();
		});
	});
	describe('onChangeTime', () => {
		it('should call onSet at time set', () => {
			const onSet = jest.fn();
			const wrapper = shallow(<TimeGroup
				onSet={onSet}
				onFocus={() => {}}
			/>);
			const { onChange } = wrapper.find('TimeField').at(0).props();
			onChange('hours', 12);
			expect(onSet).toHaveBeenCalledWith(12, 0);
		});
	});
	it('should test default idiot props ¯\\_(ツ)_/¯', () => {
		const wrapper = shallow(<TimeGroup
			onFocus={() => {}}
		/>);

		// This is just to check if default change if callable
		const { onSet, handleKeyPress } = wrapper.instance().props;
		onSet();
		handleKeyPress();	// Expect nothing to happen as it's default is () => {}
	});
});
