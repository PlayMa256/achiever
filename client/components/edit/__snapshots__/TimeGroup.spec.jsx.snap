// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeGroup render should not render if hidden 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TimeGroup
    disabled={false}
    handleKeyPress={[Function]}
    hidden={true}
    label=""
    onFocus={[Function]}
    onSet={[Function]}
    referenceHour={9}
    shouldHaveFocus={false}
    tabIndexes={0}
    time={
      Object {
        "hours": 0,
        "minutes": 0,
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): null,
  Symbol(enzyme.__nodes__): Array [
    null,
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`TimeGroup render should render two TimeFields 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TimeGroup
    disabled={false}
    handleKeyPress={[Function]}
    hidden={false}
    label=""
    onFocus={[Function]}
    onSet={[Function]}
    referenceHour={9}
    shouldHaveFocus={false}
    tabIndexes={0}
    time={
      Object {
        "hours": 0,
        "minutes": 0,
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <label
          htmlFor="12345_h"
        >
          
        </label>,
        <TimeField
          disabled={false}
          handleKeyPress={[Function]}
          mode="hours"
          onChange={[Function]}
          onFocus={[Function]}
          referenceHour={9}
          shouldHaveFocus={false}
          tabIndex={0}
          value={0}
        />,
        <span
          className="separator"
        >
          :
        </span>,
        <TimeField
          disabled={false}
          handleKeyPress={[Function]}
          mode="minutes"
          onChange={[Function]}
          onFocus={[Function]}
          referenceHour={9}
          shouldHaveFocus={false}
          tabIndex={1}
          value={0}
        />,
      ],
      "className": "time-group ",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "",
          "htmlFor": "12345_h",
        },
        "ref": null,
        "rendered": "",
        "type": "label",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "disabled": false,
          "handleKeyPress": [Function],
          "mode": "hours",
          "onChange": [Function],
          "onFocus": [Function],
          "referenceHour": 9,
          "shouldHaveFocus": false,
          "tabIndex": 0,
          "value": 0,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": ":",
          "className": "separator",
        },
        "ref": null,
        "rendered": ":",
        "type": "span",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "disabled": false,
          "handleKeyPress": [Function],
          "mode": "minutes",
          "onChange": [Function],
          "onFocus": [Function],
          "referenceHour": 9,
          "shouldHaveFocus": false,
          "tabIndex": 1,
          "value": 0,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <label
            htmlFor="12345_h"
          >
            
          </label>,
          <TimeField
            disabled={false}
            handleKeyPress={[Function]}
            mode="hours"
            onChange={[Function]}
            onFocus={[Function]}
            referenceHour={9}
            shouldHaveFocus={false}
            tabIndex={0}
            value={0}
          />,
          <span
            className="separator"
          >
            :
          </span>,
          <TimeField
            disabled={false}
            handleKeyPress={[Function]}
            mode="minutes"
            onChange={[Function]}
            onFocus={[Function]}
            referenceHour={9}
            shouldHaveFocus={false}
            tabIndex={1}
            value={0}
          />,
        ],
        "className": "time-group ",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "",
            "htmlFor": "12345_h",
          },
          "ref": null,
          "rendered": "",
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "disabled": false,
            "handleKeyPress": [Function],
            "mode": "hours",
            "onChange": [Function],
            "onFocus": [Function],
            "referenceHour": 9,
            "shouldHaveFocus": false,
            "tabIndex": 0,
            "value": 0,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": ":",
            "className": "separator",
          },
          "ref": null,
          "rendered": ":",
          "type": "span",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "disabled": false,
            "handleKeyPress": [Function],
            "mode": "minutes",
            "onChange": [Function],
            "onFocus": [Function],
            "referenceHour": 9,
            "shouldHaveFocus": false,
            "tabIndex": 1,
            "value": 0,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
